cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/timeseries/FactCodec.java

```diff
@@ -181,8 +187,12 @@ public byte[] getNextRowKey(byte[] rowKey, int indexOfDimValueToChange) {
     return newRowKey;
   }
 
-  private long roundToResolution(long ts) {
-    return (ts / resolution) * resolution;
+  public long roundToResolution(long ts, long now) {
+    long rounded = (ts / resolution) * resolution;
+    if (now - rounded > resolution * coarseLagFactor) {
+      return (rounded / resolution / coarseRoundFactor) * resolution * coarseRoundFactor;
```

Comment:
Can we use more parenthesis? To make the order of operations clears (specially with the three divides)

Reply:
What parenthesis do you propose? I tried to make it clear we round by a bigger value (resolution times coarseRoundFactor). To make it more clear, I can extract this value as a separate var. Let me do it